discovery/sync_manager.go

```diff
@@ -0,0 +1,619 @@
+	// necessary, a sync transition to an activeSync sync type will be
+	// performed to ensure we receive new graph updates from this peer.
+	startSyncer := func(s *activeSyncer) error {
+		if s.transition {
+			err := s.ProcessSyncTransition(activeSync)
+			if err != nil {
+				return fmt.Errorf("unable to transition to "+
+					"%v: %v", activeSync, err)
+			}
+			s.transition = false
+
+			m.syncersMtx.Lock()
+			m.activeSyncers[s.cfg.peerPub] = s.gossipSyncer
+			delete(m.syncers, s.cfg.peerPub)
+			m.syncersMtx.Unlock()
+		}
+
+		log.Debugf("Starting active syncer for peer=%x", s.cfg.peerPub)
+
+		startNext = s.ResetSyncedSignal()
+		m.cfg.ActiveSyncerTimeoutTicker.Resume()
+		s.Start()
+
+		return nil
+	}
+
+	// startNextSyncer is a helper closure that we'll use to start the next
+	// syncer queued up. If there aren't any, this will act as a NOP.
+	startNextSyncer := func() {
+		current = m.nextPendingActiveSyncer()
+		for current != nil {
+			err := startSyncer(current)
+			if err == nil {
+				return
+			}
+
+			log.Errorf("Unable to start gossipSyncer(%x): %v",
+				current.cfg.peerPub, err)
+
+			current = m.nextPendingActiveSyncer()
+		}
+
+		startNext = nil
+		m.cfg.ActiveSyncerTimeoutTicker.Pause()
+	}
+
+	for {
+		select {
+		// Our RotateTicker has ticked, so we'll attempt to rotate any
+		// of our activeSyncers with randomly chosen ones.
+		case <-m.cfg.RotateTicker.Ticks():
+			m.rotateActiveSyncers()
+
+		// Our FullSyncTicker has ticked, so we'll randomly select a
+		// peer and force a full sync with them.
+		case <-m.cfg.FullSyncTicker.Ticks():
+			m.forceFullSync()
+
+		// A new activeSyncer has been received, so we'll either attempt
+		// to start it if the queue is empty or queue it so that we
+		// don't attempt to sync with multiple peers at the same time.
+		case s := <-m.newActiveSyncers:
+			if current == nil {
+				if err := startSyncer(s); err != nil {
+					log.Errorf("Unable to start "+
+						"gossipSyncer(%x): %v", err)
+					continue
+				}
+
+				current = s
+				continue
+			}
+
+			log.Debugf("Queueing next active syncer for peer=%x",
+				s.cfg.peerPub)
+
+			m.pendingActiveSyncers = append(m.pendingActiveSyncers, s)
+
+		// A stale activeSyncer has been received, so we'll need to
+		// remove them from our queue. If we are currently waiting for
+		// its state machine to complete, we'll move on to the next
+		// activeSyncer in the queue.
+		case peer := <-m.staleActiveSyncers:
+			if current != nil && peer == current.cfg.peerPub {
+				startNextSyncer()
+				continue
+			}
+
+			for i, activeSyncer := range m.pendingActiveSyncers {
+				if activeSyncer.cfg.peerPub == peer {
+					m.pendingActiveSyncers = append(
+						m.pendingActiveSyncers[:i],
+						m.pendingActiveSyncers[i+1:]...,
+					)
+					break
+				}
+			}
+
+		// Our current active syncer has reached its terminal
+		// chansSynced state, so we'll proceed to starting the next
+		// pending active syncer if there is one.
+		case <-startNext:
```

Comment:
Why do we care if active syncers reach a synced state? They should start in this state more or less, since passive->active is just a matter of setting a new gossip timestamp.

Reply:
> Why do we care if active syncers reach a synced state?

This is how the round-robin works to ensure we only have one syncer negotiating at a time.

> They should start in this state more or less, since passive->active is just a matter of setting a new gossip timestamp.

`passiveSync` -> `activeSync` also triggers the state machine to start from `syncingChans`.

Reply:
Couldn't it be a problem if we start a new syncer that (almost) immediately goes to `chansSynced`, triggering us to start a new one, same thing happens again etc...? Possibly need an active syncer stay our chosen one for a while.

Reply:
I don't see a problem with that. The round-robin is just to ensure that we do the sync exchange with one active syncer at a time. New graph updates can be received from any active syncer, regardless of their order in the round-robin.

Reply:
Imagine there are no new updates, and all nodes are perfectly in sync. The problem now is that you'll choose an active syncer, but since you are already in sync you'll immediately move to `chansSynced` and trigger the `startNext` signal. Now you'll continue rapidly switching active syncers. Or is there something here that prevents that scenario?

Reply:
Again, the round-robin is just to ensure we only query one peer at a time for _historical_ updates. Once a syncer reaches `chansSynced`, we can receive _new_ updates from the peer at any point.